### PR TITLE
Issues/188 path ends with

### PR DESCRIPTION
**Pull request by gaconkzk**
Implement Path.endWiths ( #188 )
----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt [hunk 1]
```diff
@@ -33,6 +33,18 @@ fun <T : Path> Expect<T>.exists(): Expect<T> = addAssertion(ExpectImpl.path.exis
  */
 fun <T : Path> Expect<T>.existsNot(): Expect<T> = addAssertion(ExpectImpl.path.existsNot(this))
 
+
+/**
+ * Expects that the subject of the assertion (a [Path]) is ends with the other [Path];
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt]:
```suggestion
 * Expects that the subject of the assertion (a [Path]) ends with the expected [Path];
```

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt [hunk 2]
```diff
@@ -33,6 +33,18 @@ fun <T : Path> Expect<T>.exists(): Expect<T> = addAssertion(ExpectImpl.path.exis
...
+ *
+ * @param other The [Path] provided to the assertion
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt]:
```suggestion
 * @param expected The [Path] provided to the assertion
```

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt [hunk 3]
```diff
@@ -33,6 +33,18 @@ fun <T : Path> Expect<T>.exists(): Expect<T> = addAssertion(ExpectImpl.path.exis
...
+ * @return This assertion container to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ *
+ * @since 0.9.0
+ */
+fun <T : Path> Expect<T>.endsWith(other: T): Expect<T> = addAssertion(ExpectImpl.path.endsWith(this, other))
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/pathAssertions.kt]: We always use `expected`, let's stick to this wording
```suggestion
fun <T : Path> Expect<T>.endsWith(expected: T): Expect<T> = addAssertion(ExpectImpl.path.endsWith(this, expected))
```

File: domain/api/atrium-domain-api-jvm/src/main/kotlin/ch/tutteli/atrium/domain/creating/PathAssertions.kt [hunk 4]
```diff
@@ -20,5 +20,6 @@ val pathAssertions by lazy { loadSingleService(PathAssertions::class) }
 interface PathAssertions {
     fun <T : Path> exists(assertionContainer: Expect<T>): Assertion
     fun <T : Path> existsNot(assertionContainer: Expect<T>): Assertion
+    fun <T : Path> endsWith(assertionContainer: Expect<T>, other: T): Assertion
```
Comment (robstoll) [inline domain/api/atrium-domain-api-jvm/src/main/kotlin/ch/tutteli/atrium/domain/creating/PathAssertions.kt]:
```suggestion
    fun <T : Path> endsWith(assertionContainer: Expect<T>, expected: T): Assertion
```

File: domain/builders/atrium-domain-builders-jvm/src/main/kotlin/ch/tutteli/atrium/domain/builders/creating/PathAssertionsBuilder.kt [hunk 5]
```diff
@@ -20,6 +20,9 @@ object PathAssertionsBuilder : PathAssertions {
     override inline fun <T : Path> existsNot(assertionContainer: Expect<T>) =
         pathAssertions.existsNot(assertionContainer)
 
+    override fun <T : Path> endsWith(assertionContainer: Expect<T>, other: T) =
```
Comment (robstoll) [inline domain/builders/atrium-domain-builders-jvm/src/main/kotlin/ch/tutteli/atrium/domain/builders/creating/PathAssertionsBuilder.kt]:
```suggestion
    override fun <T : Path> endsWith(assertionContainer: Expect<T>, expected: T) =
```

File: domain/builders/atrium-domain-builders-jvm/src/main/kotlin/ch/tutteli/atrium/domain/builders/creating/PathAssertionsBuilder.kt [hunk 6]
```diff
@@ -20,6 +20,9 @@ object PathAssertionsBuilder : PathAssertions {
...
+        pathAssertions.endsWith(assertionContainer, other)
```
Comment (robstoll) [inline domain/builders/atrium-domain-builders-jvm/src/main/kotlin/ch/tutteli/atrium/domain/builders/creating/PathAssertionsBuilder.kt]:
```suggestion
        pathAssertions.endsWith(assertionContainer, expected)
```

File: domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/pathAssertions.kt [hunk 7]
```diff
@@ -20,6 +21,9 @@ fun <T : Path> _exists(assertionContainer: Expect<T>): Assertion =
 fun <T : Path> _existsNot(assertionContainer: Expect<T>): Assertion =
     ExpectImpl.builder.createDescriptive(assertionContainer, NOT_TO, RawString.create(EXIST)) { it.notExists }
 
+fun <T : Path> _endsWith(assertionContainer: Expect<T>, other: T): Assertion =
+    ExpectImpl.builder.createDescriptive(assertionContainer, TO, RawString.create(ENDS_WITH)) { it.endsWith(other) }
```
Comment (robstoll) [inline domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/pathAssertions.kt]: that's not what you want. Make a test fail on purpose and have a look how it looks like

File: domain/robstoll/atrium-domain-robstoll-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/creating/PathAssertionsImpl.kt [hunk 8]
```diff
@@ -10,6 +11,7 @@ import java.nio.file.Path
 class PathAssertionsImpl : PathAssertions {
     override fun <T : Path> exists(assertionContainer: Expect<T>) = _exists(assertionContainer)
     override fun <T : Path> existsNot(assertionContainer: Expect<T>) = _existsNot(assertionContainer)
+    override fun <T : Path> endsWith(assertionContainer: Expect<T>, other: T) = _endsWith(assertionContainer, other)
```
Comment (robstoll) [inline domain/robstoll/atrium-domain-robstoll-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/creating/PathAssertionsImpl.kt]:
```suggestion
    override fun <T : Path> endsWith(assertionContainer: Expect<T>, expected: T) = _endsWith(assertionContainer, expected)
```

File: domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/pathAssertions.kt (hunk 7 above)
Comment (gaconkzk) [reply domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/pathAssertions.kt]: Thanks. I changed the descriptive to match with requirement

**Comment by robstoll** [comment]
@gaconkzk please rebase on master as there is a conflict

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt [hunk 9]
```diff
@@ -2,12 +2,14 @@ package ch.tutteli.atrium.api.fluent.en_GB.jdk8
...
 import java.nio.file.Path
 
 class PathAssertionsSpec : ch.tutteli.atrium.specs.integration.PathAssertionsSpec(
     fun0(Expect<Path>::exists),
-    fun0(Expect<Path>::existsNot)
+    fun0(Expect<Path>::existsNot),
+    fun1(Expect<Path>::endsWith)
 ){
     @Suppress("unused", "UNUSED_VALUE")
     private fun ambiguityTest() {
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: please add endsWith to the ambiguityTest

Comment (robstoll) [reply apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: I'll fix it myself

**Comment by robstoll** [comment]
@gaconkzk thanks for the contribution :+1: 

Fancy another one?
https://github.com/robstoll/atrium/issues?q=is%3Aissue+is%3Aopen+label%3Ahacktoberfest+no%3Aassignee

(and don't forget to star Atrium if you liked it :wink:)

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt (hunk 9 above)
Comment (gaconkzk) [reply apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: I'm not sure about the purpose, but are you want something like those in the `ambiguityTest()` function

```kotlin

        val a3: Expect<Path> = notImplemented()
        val a4: Path = notImplemented()

        a3.endsWith(a4)
```

Comment (robstoll) [reply apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/PathAssertionsSpec.kt]: The purpose is to see whether the signature of the method makes sense (as this is sometimes hard to see in the review). Your current implementation for instance fails the ambiguityTest:
```
a1.endsWith(Paths.get("a"))
a2.endsWith(Paths.get("a"))
```
But no worries I'll fix it myself unless you would like to fix it?

**Pull request merged**